### PR TITLE
[WIP] Use `cudaOccupancyMaxPotentialBlockSize` to determine kernel size

### PR DESCRIPTION
**Pull request by asi1024**
Ref #2731.

### CODE REVIEW

File: cupy/cuda/function.pyx [hunk 1]
```diff
@@ -165,15 +166,21 @@ cdef class Function:
             max(1, block[0]), max(1, block[1]), max(1, block[2]),
             args, shared_mem, s)
 
+    cpdef launch(
+            self, args, size_t gridx, size_t blockx, size_t shared_mem=0,
+            stream=None):
+        _launch(
+            self.ptr, gridx, 1, 1, blockx, 1, 1,
```
Comment (niboshi) [inline cupy/cuda/function.pyx]: Do you think we don't have to expose y and z as well?

**Comment by niboshi** [comment]
Please post some benchmark.

**Comment by asi1024** [comment]
This PR does not affect performance. This PR simplifies the return value of `_get_block_specs`, because new `linear_launch` does not require `size` parameter.

master (`b11695907`)
```
cupy_add            :    10.701 us   +/- 0.388 (min:   10.438 / max:   22.952) us     14.640 us   +/- 0.452 (min:   12.896 / max:   26.912) us
cupy_add            :    11.298 us   +/- 0.274 (min:   11.068 / max:   18.679) us     15.222 us   +/- 0.375 (min:   13.952 / max:   22.720) us
cupy_add            :    11.135 us   +/- 2.259 (min:   10.647 / max:  130.486) us    552.666 us   +/- 2.478 (min:  549.504 / max:  675.168) us
cupy_add            :    11.696 us   +/- 2.495 (min:   11.228 / max:  146.393) us    488.504 us   +/-51.735 (min:  483.840 / max: 3313.440) us
sum                 :    14.549 us   +/- 3.383 (min:   14.056 / max:  198.808) us     18.511 us   +/- 3.528 (min:   16.928 / max:  210.048) us
sum                 :    14.796 us   +/- 0.380 (min:   14.424 / max:   29.508) us     18.783 us   +/- 0.471 (min:   17.184 / max:   33.440) us
sum                 :    15.050 us   +/- 0.396 (min:   14.506 / max:   21.618) us    202.969 us   +/- 0.611 (min:  200.928 / max:  209.344) us
sum                 :    15.402 us   +/- 0.422 (min:   14.804 / max:   23.510) us    177.745 us   +/- 1.188 (min:  174.624 / max:  184.512) us
```

This PR (`c7a4418cc`)
```
cupy_add            :    10.576 us   +/- 0.185 (min:   10.375 / max:   15.050) us     14.177 us   +/- 0.333 (min:   12.544 / max:   22.144) us
cupy_add            :    11.230 us   +/- 0.256 (min:   10.997 / max:   17.080) us     14.829 us   +/- 0.323 (min:   13.600 / max:   20.832) us
cupy_add            :    10.970 us   +/- 2.378 (min:   10.498 / max:  137.158) us    552.096 us   +/- 2.580 (min:  547.712 / max:  680.608) us
cupy_add            :    11.639 us   +/- 3.620 (min:   11.142 / max:  150.669) us    487.115 us   +/- 3.738 (min:  484.928 / max:  630.848) us
sum                 :    14.231 us   +/- 0.403 (min:   13.825 / max:   24.457) us     17.864 us   +/- 0.516 (min:   16.544 / max:   28.576) us
sum                 :    14.517 us   +/- 0.409 (min:   14.164 / max:   28.130) us     18.225 us   +/- 0.755 (min:   14.464 / max:   32.352) us
sum                 :    14.719 us   +/- 3.289 (min:   14.093 / max:  193.734) us    202.590 us   +/- 3.425 (min:  200.800 / max:  387.456) us
sum                 :    15.106 us   +/- 0.477 (min:   14.577 / max:   23.760) us    177.484 us   +/- 1.249 (min:  174.080 / max:  187.072) us
```

File: cupy/cuda/function.pyx (hunk 1 above)
Comment (asi1024) [reply cupy/cuda/function.pyx]: We can specify them by using `Funciton.__call__`.

**Comment by asi1024** [comment]
This PR is not so effective for the performance. (see https://github.com/cupy/cupy/pull/2827#issuecomment-566973948)

**Comment by leofang** [comment]
It could be that element wise operations are memory bound, so max occupancy would increase too many memory transactions and/or too much register pressure. Down-tune a bit the block size might help. Another potential factor to look at is to check data alignment.

**Pull request closed without merging**